docs/history-service.md

```diff
@@ -0,0 +1,274 @@
+
+- **Scaling**: the History service must be able to support arbitrarily many concurrent workflow executions. See [History shards](#history-shards).
+- **Consistency**: the Workflow and Activity tasks in Matching Service must be consistent with the workflow execution data in History Service. See [Consistency guarantees](#consistency-guarantees).
+- **Latency** of request handling and server-side workflow advance must be minimized.
+
+The rest of this document describes the implementation, which addresses these requirements.
+
+The History Service has many other responsibilities which are not covered here, for example: Shard Membership and Ownership, Replication and conflict resolution, Visibility, Archival, Workflow Versioning, Workflow Reset, Scheduled workflows.
+
+## Concepts
+
+### History shards
+
+A single Temporal Cluster manages many (e.g. millions of) Workflow Executions.
+These are logically partitioned into subsets that we call History Shards.
+The Temporal Cluster typically has multiple instances of the History service application process (each typically running on a separate machine/VM), and each History service instance owns some set of History Shards.
+"Owning" a History Shard means being responsible for the lifecycle of every workflow execution in the shard by [handling incoming requests](#rpc-handling) relating to that workflow execution, managing timers, and causing Workflow and Activity Tasks to be enqueued in Matching Service. To achieve this, each shard manages its own [task queues](#queue-processing).
+
+Therefore, objects corresponding to History Shards exist in the memory of a History service instance.
+Persistence details are abstracted away via a Go interface but in practice, when the persistence layer is backed by Cassandra, History Shards map 1-1 with partitions of the [`executions` table](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/schema/cassandra/temporal/schema.cql#L52). Persistence layer implementations also exist for MySQL, Postgres, and SQLite.
+
+<details>
+<summary><i>Code entrypoints</i></summary>
+
+- [`schema/`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/schema) directory
+- When the History service starts on a host, [`Start()`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/history_engine.go#L288) is called for each shard owned by that host.
+
+</details>
+
+### Workflow Execution "Mutable State"
+
+For every workflow execution, we maintain a collection of data structures summarizing various aspects of its current state, for example, the identities of in-progress activities, timers, and child workflows.
+Although these data could in principle be recomputed from Workflow History Events when handling an incoming request, this would be slow, and hence the summaries themselves are persisted.
+
```

Comment:
maybe say somewhere that some number of mutable state are cached in memory?

Reply:
Thanks, added.